src/Interpreters/Cache/FileCache.cpp

```diff
@@ -1414,45 +1414,69 @@ void FileCache::applySettingsIfPossible(const FileCacheSettings & new_settings,
-                main_priority->modifySizeLimits(
-                    new_settings.max_size, new_settings.max_elements,
-                    new_settings.slru_size_ratio, cache_lock);
+                if (eviction_candidates.size() == 0)
+                {
+                    main_priority->modifySizeLimits(
+                        new_settings.max_size, new_settings.max_elements,
+                        new_settings.slru_size_ratio, cache_lock);
+                }
+                else
+                {
+                    /// Remove only queue entries of eviction candidates.
+                    eviction_candidates.removeQueueEntries(cache_lock);
+                    /// Note that (in-memory) metadata about corresponding file segments
+                    /// (e.g. file segment info in CacheMetadata) will be removed
+                    /// only after eviction from filesystem. This is needed to avoid
+                    /// a race on removal of file from filesystsem and
+                    /// addition of the same file as part of a newly cached file segment.
+
+                    /// Modify cache size limits.
+                    /// From this point cache eviction will follow them.
+                    main_priority->modifySizeLimits(
+                        new_settings.max_size, new_settings.max_elements,
+                        new_settings.slru_size_ratio, cache_lock);
+
+                    cache_lock.unlock();
+
+                    auto finalize_eviction = [&]()
+                    {
+                        if (eviction_candidates.requiresLockToFinalize())
+                        {
+                            cache_lock.lock();
+                            eviction_candidates.finalize(nullptr, &cache_lock);
+                        }
+                        else
+                            eviction_candidates.finalize(nullptr);
+                    };
+
+                    try
+                    {
+                        /// Do actual eviction from filesystem.
+                        eviction_candidates.evict();
+                    }
+                    catch (...)
+                    {
+                        tryLogCurrentException(__PRETTY_FUNCTION__);
+                        finalize_eviction();
+                        throw;
+                    }
+
+                    finalize_eviction();
+                }
 
                 modified_size_limit = true;
             }
         }
 
         if (modified_size_limit)
         {
-            try
-            {
-                /// Do actual eviction from filesystem.
-                eviction_candidates.evict();
-            }
-            catch (...)
-            {
-                if (eviction_candidates.needFinalize())
-                    eviction_candidates.finalize(nullptr, lockCache());
-                throw;
-            }
-
-            if (eviction_candidates.needFinalize())
-                eviction_candidates.finalize(nullptr, lockCache());
```

Comment:
Second fix -- we did not always call finalize, but we should have because we needed to release space holders